crypto/params_from_text.c

```diff
@@ -62,20 +63,25 @@ static int prepare_from_text(const OSSL_PARAM *paramdefs, const char *key,
          * buffer, i.e. if it's negative, we need to deal with it.  We do
          * it by subtracting 1 here and inverting the bytes in
          * construct_from_text() below.
+         * To subtract 1 from an absolute value of a negative number we
+         * actually have to add 1: -3 - 1 = -4, |-3| = 3 + 1 = 4.
          */
         if (p->data_type == OSSL_PARAM_INTEGER && BN_is_negative(*tmpbn)
-            && !BN_sub_word(*tmpbn, 1)) {
+            && !BN_add_word(*tmpbn, 1)) {
             return 0;
         }
 
-        *buf_n = BN_num_bytes(*tmpbn);
+        buf_bits = (size_t)BN_num_bits(*tmpbn);
+        *buf_n = (buf_bits + 7) / 8;
```

Comment:
Can you explain why you need to know the amount of bits, especially since you're multiplying with 8 whenever you compare with it?

Reply:
I need to know when an `OSSL_PARAM_INTEGER` has less than the maximum number of bits. This allows `INT32_MAX` to fit into `int_32t`, but raises an error when someone wants to put `INT32_MAX + 1` there. In other words, the number of bits is significant to determine whether a signed value can fit into an integer of given byte-size, or not.
```
            if (buf_bits > p->data_size * 8
                || (p->data_type == OSSL_PARAM_INTEGER
                    && buf_bits == p->data_size * 8)) {
```

Reply:
Ah, yeahok, that makes sense